[WIP] Add NucleusX Model

syncdoth opened the pull request:
# What does this PR do?

<!--
Congratulations! You've made it this far! You're not quite done yet though.

Once merged, your PR is going to appear in the release notes with the title you set, so make sure it's a great title that fully reflects the extent of your awesome contribution.

Then, please replace this with a description of the change and which issue is fixed (if applicable). Please also include relevant motivation and context. List any dependencies (if any) that are required for this change.

Once you're done, someone will review your PR shortly (see the section "Who can review?" below to tag some potential reviewers). They may suggest changes to make the code even better. If no one reviewed your PR after a week has passed, don't hesitate to post a new comment @-mentioning the same persons---sometimes notifications get lost.
-->

<!-- Remove if not applicable -->

Add a new model named NucleusX Model.


## Before submitting
- [ ] This PR fixes a typo or improves the docs (you can dismiss the other checks if that's the case).
- [x] Did you read the [contributor guideline](https://github.com/huggingface/transformers/blob/main/CONTRIBUTING.md#start-contributing-pull-requests),
      Pull Request section?
- [ ] Was this discussed/approved via a Github issue or the [forum](https://discuss.huggingface.co/)? Please add a link
      to it if that's the case.
- [x] Did you make sure to update the documentation with your changes? Here are the
      [documentation guidelines](https://github.com/huggingface/transformers/tree/main/docs), and
      [here are tips on formatting docstrings](https://github.com/huggingface/transformers/tree/main/docs#writing-source-documentation).
- [x] Did you write any new necessary tests?


## Who can review?

Anyone in the community is free to review the PR once the tests have passed. Feel free to tag
members/contributors who may be interested in your PR.

<!-- Your PR will be replied to more quickly if you can figure out the right person to tag with @

 If you know how to use git blame, that is the easiest way, otherwise, here is a rough guide of **who to tag**.
 Please tag fewer than 3 people.

Models:

- text models: @ArthurZucker and @younesbelkada
- vision models: @amyeroberts
- speech models: @sanchit-gandhi
- graph models: @clefourrier

Library:

- flax: @sanchit-gandhi
- generate: @gante
- pipelines: @Narsil
- tensorflow: @gante and @Rocketknight1
- tokenizers: @ArthurZucker
- trainer: @muellerzr and @pacman100

Integrations:

- deepspeed: HF Trainer/Accelerate: @pacman100
- ray/raytune: @richardliaw, @amogkam
- Big Model Inference: @SunMarc
- quantization (bitsandbytes, autogpt): @SunMarc and @younesbelkada

Documentation: @stevhliu and @MKhalusova

HF projects:

- accelerate: [different repo](https://github.com/huggingface/accelerate)
- datasets: [different repo](https://github.com/huggingface/datasets)
- diffusers: [different repo](https://github.com/huggingface/diffusers)
- rust tokenizers: [different repo](https://github.com/huggingface/tokenizers)

Maintained examples (not research project or legacy):

- Flax: @sanchit-gandhi
- PyTorch: See Models above and tag the person corresponding to the modality of the example.
- TensorFlow: @Rocketknight1

 -->

Commit:
[WIP] Add NucleusX Model

Commit:
solve chunkwise equivalence and add test

Commit:
ignore attention checking for nucleusx

Commit:
update license

Commit:
implement generate from input_embeds

Commit:
handle weight initialization issue

Commit:
update NucleusX docs

syncdoth commented on tests/models/nucleus_x/test_modeling_nucleus_x.py:
```diff
@@ -0,0 +1,397 @@
+        config_and_inputs = self.model_tester.prepare_config_and_inputs()
+        self.model_tester.create_and_check_model(*config_and_inputs)
+
+    def test_model_various_embeddings(self):
```
delete this test case.

sippycoder commented on src/transformers/models/nucleus_x/configuration_nucleus_x.py:
```diff
@@ -0,0 +1,215 @@
+            relevant if `config.is_decoder=True`.
+        forward_mode (`str`, *optional*, defaults to `"parallel"`):
+            The three modes of Retentive Network: `"parallel"`, `"recurrent"`, or `"chunkwise"`.
+        activation_fn (`str` or `function`, *optional*, defaults to `"gelu"`):
```
We default to `swish` activation.

sippycoder commented on src/transformers/models/nucleus_x/configuration_nucleus_x.py:
```diff
@@ -0,0 +1,215 @@
+
+    def __init__(
+        self,
+        vocab_size: int = 32000,
```
Are we sure we pre-trained our models with `vocab_size = 32000`? I think we have a bug with `50257` in our pre-training. Can you check this?

syncdoth replied:
In our train code, we reset the vocab size based on the tokenizer we use. So our models will have vocab size of 32000 (or 32001 for 7b, including pad_token. For 13b, I'm using unk_token as pad_token following openllama.)

sippycoder commented on src/transformers/models/nucleus_x/modeling_nucleus_x.py:
```diff
@@ -0,0 +1,1410 @@
+            retention_rel_pos = (
+                (sin, cos),
+                (inner_mask, cross_decay, query_inner_decay, value_inner_decay, decay_scale),
+            )
```
Nothing major here. Please move this block to a different function. Long functions are less readable.

sippycoder commented on src/transformers/models/nucleus_x/modeling_nucleus_x.py:
```diff
@@ -0,0 +1,1410 @@
+            past_key_value = past_key_values[idx] if past_key_values is not None else None
+
+            if self.gradient_checkpointing and self.training:
+
```
Please add a logic to make sure use_cache is `False`
Ref: https://github.com/huggingface/transformers/blob/5125fb9aa0803580995445561c1d5375f5f79c56/src/transformers/models/falcon/modeling_falcon.py#L776

Commit:
fix typo; init rmsnorm in _init_weights

sippycoder commented on src/transformers/models/nucleus_x/modeling_nucleus_x.py:
```diff
@@ -0,0 +1,1410 @@
+        retention = retention * decay_mask
+
+        # invariant after normalization
+        retention = retention / retention.detach().sum(dim=-1, keepdim=True).abs().clamp(min=1)
```
Is `detach()` necessary here?? Is there a solid reason why this would make sense?

sippycoder replied:
I see that this is used in the torch scale implementation as well. So weird.

syncdoth replied:
I have no solid idea why we need detach. But I guess this here is for pure scaling purposes for numerical stability, so they don't extra want gradient operations here.

sippycoder commented on src/transformers/models/nucleus_x/modeling_nucleus_x.py:
```diff
@@ -0,0 +1,1410 @@
+
+        qk_mat = q @ k_t  # [b, n_c, h, t_c, t_c]
+        qk_mat = qk_mat * decay_mask.unsqueeze(1)
+        inner_scale = qk_mat.detach().abs().sum(dim=-1, keepdim=True).clamp(min=1)
```
Why does using `detach()` make sense?

sippycoder commented on src/transformers/models/nucleus_x/modeling_nucleus_x.py:
```diff
@@ -0,0 +1,1410 @@
+            kv_recurrent.append(kv_state / kv_scale)
+            cross_scale.append(kv_scale)
+            kv_state = kv_state * cross_decay + kv[:, i]
+            kv_scale = kv_state.detach().abs().sum(dim=-2, keepdim=True).max(dim=-1, keepdim=True).values.clamp(min=1)
```
`detach()` is used here again.

sippycoder reviewed (commented):
About 60% reviewed. Will continue.

Commit:
refactor relpos, check use_cache with gradient_checkpoint, default activation_fn=swish

Commit:
necessary changes for code styling

Commit:
fix nucleusai url

Commit:
remove nucleus_x._set_gradient_checkpointing

Commit:
remove timm dependency and copy drop_path

Commit:
resolve unused config, init parameter of NucleusXModels

Commit:
add reset_parameters to NucleusXRMSNorm

Commit:
refactor past_key_value to tuple, update docs and tests accordingly

Commit:
refactor argument docs, adds to _toctree

Commit:
add empty returns

Commit:
put nucleus_x under text model in _toctree

Commit:
update documents

Commit:
revert added comma

Removes the accidentally added comma in tests/generation/test_utils.py

Commit:
better handling of dtype for half precision weights

When loading the model weights in different dtype other than fp32,
`.float` statements may cause troubles. This commit handles the tensor
creations and float casting to be aware of the `dtype` of the weights.

Commit:
rename layernorm -> rms_norm

Commit:
remove subln

Reason: since we are using GLU, sub-layernorm is not well-defined.

Commit:
test parallel vs recurrent with 10 random examples

Commit:
remove build_rms_norm, redefine NucleusXRMSNorm

This follows the example of other models, such as LongT5, idefics, llama, etc.

Commit:
fix typo, inline init of ffn, smaller example model size

This resolves the comments by @Rocketknight1

Commit:
update docs and checkpoint ref

Commit:
integration testing
